### PR TITLE
Allow to specify template when running `dagger project init`

### PR DESCRIPTION
**Pull request by marcosnils**
Adds the ability to select a template which will create a new file
in the CWD with the template name. Templates present in the
"cmd/dagger/project/templates" directory are automatically embedded in the dagger
binary when building and then listed in the `dagger project init` help output.

Usage:

    dagger project init -t <template_name>

One idea that we had while pairing on this is to eventually try
improving this UX by extending the `dagger project` command by adding
sub-commands like `dagger project templates [list, new, output, ...]`.

Pair: gerhard

Signed-off-by: Marcos Lilljedahl <marcosnils@gmail.com>

### CODE REVIEW

**Comment by marcosnils** [comment]
cc @shykes and @helderco

**Comment by TomChv** [comment]
Related to #1716

File: cmd/dagger/cmd/project/templates/hello.cue
Comment (TomChv) [inline cmd/dagger/cmd/project/templates/hello.cue:13]: Since we want to incite people to use universe package and only `core` when it's necessary, I think it's better to use universe package in `hello`?

So you could update `core.#Pull` with `docker.#Pull` or `alpine.#Build` and `core.#Exec` with `bash.#Run` or `docker.#Run`.

File: tests/project.bats
Comment (TomChv) [inline tests/project.bats:15]: Why duplicate?

File: tests/project.bats
Comment (TomChv) [inline tests/project.bats:53]: `exit 1` instead of `return`?

File: tests/project.bats
Comment (TomChv) [inline tests/project.bats:63]: `exit 1` instead of `return`?

**Review by TomChv — COMMENTED** [review]
Nice contribution, I really like that one!
I left few comments to improve the PR

File: tests/project.bats
Comment (marcosnils) [reply tests/project.bats:15]: Missd this when rebasing conflicts, thx for :eyes:

File: tests/project.bats
Comment (marcosnils) [reply tests/project.bats:63]: +1 changing

File: tests/project.bats
Comment (marcosnils) [reply tests/project.bats:53]: +1 chaning

File: cmd/dagger/cmd/project/templates/hello.cue
Comment (helderco) [reply cmd/dagger/cmd/project/templates/hello.cue:19]: It's on the package style guide: https://docs.dagger.io/1226/coding-style#use-top-to-match-anything

File: cmd/dagger/cmd/project/template.go
Comment (marcosnils) [reply cmd/dagger/cmd/project/template.go:39]: I've seen this differently in multiple codebases (i.e: https://cs.opensource.google/go/go/+/master:src/os/dir.go;l=75?q=%22%5B%5Dstring%7B%7D%22&ss=go%2Fgo&start=31) 

The main reason here to return `[]string{}` is that otherwise, I'll have to either initialize it below even for empty results as we shouldn't return a `nil` value here.

File: cmd/dagger/cmd/project/template.go
Comment (helderco) [reply cmd/dagger/cmd/project/template.go:13]: I agree.

Comment (TomChv) [reply cmd/dagger/cmd/project/template.go:13]: Agreed to

File: cmd/dagger/cmd/project/templates/hello.cue
Comment (TomChv) [inline cmd/dagger/cmd/project/templates/hello.cue:27]: Why `always: true`?

File: cmd/dagger/cmd/project/templates/hello.cue
Comment (TomChv) [inline cmd/dagger/cmd/project/templates/hello.cue:18]: Could be directory: `packages: bash: _`.

Btw I've seen a lot of user using that pattern for running bash, I think we should improve `bash.#Run` so there is a default `bash` installed if user do not provide his image.
/cc @helderco

**Review by TomChv — COMMENTED** [review]
Great improvement, I left few more comments but it will be mergeable soon :rocket:

**Comment by aluzzardi** [comment]
This is great! Nice improvement.

Open questions & suggestions (should not hold back merging this, feel free to ignore):

- When initializing with a template, should we install the dependencies automatically as well? /cc @shykes
- Would it make sense to put templates in their own directory, rather than one per file? (e.g. `templates/<name>/` instead of `templates/<name>.cue`)
  - They could contain multiple files
  - We could have templates using the same file name (e.g. `main.cue`)
  - It would decouple CUE file name from template name (e.g. `default` template)
- Should we have a default template with `project init`?

File: cmd/dagger/cmd/project/templates/hello.cue
Comment (aluzzardi) [reply cmd/dagger/cmd/project/templates/hello.cue:18]: Ugh ... this is my bad, I think I broke the CUE linter when converting it to dagger (`dagger do lint cue`).

The linter should have caught this:

```
$ make cuefmt
$ git diff
diff --git a/cmd/dagger/cmd/project/templates/hello.cue b/cmd/dagger/cmd/project/templates/hello.cue
index f0b45533..5d3fdab1 100644
--- a/cmd/dagger/cmd/project/templates/hello.cue
+++ b/cmd/dagger/cmd/project/templates/hello.cue
@@ -15,9 +15,7 @@ import (
 dagger.#Plan & {
        actions: {
                _alpine: alpine.#Build & {
-                       packages: {
-                               bash: _
-                       }
+                       packages: bash: _
                }
```

Comment (aluzzardi) [reply cmd/dagger/cmd/project/templates/hello.cue:18]: Fixed the linter: https://github.com/dagger/dagger/pull/2159

@marcosnils Can you rebase this PR after #2159 is merged? Should surface CUE linting errors

**Comment by marcosnils** [comment]
> * When initializing with a template, should we install the dependencies automatically as well? /cc @shykes

We thought about this with @gerhard and we decided not to do it to keep the `project init` UX consistent. Installing dependencies when some flags (like templates) are used could be a bit misleading



> * Would it make sense to put templates in their own directory, rather than one per file? (e.g. `templates/<name>/` instead of `templates/<name>.cue`)

Good call. I guess we can start small and if we have a use-case to use multiple files we can change this? Since it won't break the API, not sure if we need this now

> * Should we have a default template with `project init`?

Good question. I'll defer this to @shykes

File: cmd/dagger/cmd/project/templates/hello.cue
Comment (marcosnils) [reply cmd/dagger/cmd/project/templates/hello.cue]: done!

File: cmd/dagger/cmd/project/templates/hello.cue
Comment (TomChv) [reply cmd/dagger/cmd/project/templates/hello.cue:25]: I agree, let's go for `always: true`

**Comment by TomChv** [comment]
> Would it make sense to put templates in their own directory, rather than one per file? (e.g. templates/<name>/ instead of templates/<name>.cue)

That's a really nice call, may we can open an issue to not lose track of this?

**Review by TomChv — APPROVED** [review]

**Comment by marcosnils** [comment]
> * Having a default template at `init`: good question, I think it would be helpful, yes. What does everyone else think?

I guess this depends on the template? I don't think a hello-world template is very much useful. Maybe a more comprehensive template that builds a docker image could be a better fit for a new project. 

> Strong preference for keeping templates in universe, like everything else. universe.dagger.io/templates/xxx is fine I think.

It's hard to make this work using `go embed` since embedding only works with files at same lever or deeper in the tree relatively to the source file that does the embedding. Given this limitation, and as an initial approach, I've opted for not adding these templates in `universe`. We can definitely find a way through dagger actions / `go generate` or some other process around the native `go build` process, but I was trying to avoid that before having to make that decision (no is temporary, yes is forever :D)

**Review by aluzzardi — APPROVED** [review]
Moving this along -- there's a few open questions (default template, moving to universe, etc) but I don't want to hold this back, we can improve in a follow up.

As @marcosnils pointed out there's no easy way right now to have this in universe. I hope we could revamp this when we rework package management and externalize `universe` (e.g. templates could be fetched over git, same as universe itself)

**Pull request merged**
